libcoap/src/oscore.rs

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright © The libcoap-rs Contributors, all rights reserved.
+ * This file is part of the libcoap-rs project, see the README file for
+ * general information on this project and the NOTICE.md and LICENSE files
+ * for information regarding copyright ownership and terms of use.
+ *
+ * oscore.rs - Wrapper for libcoap OSCORE functionality.
+ */
+
+use core::{ffi::c_void, ptr};
+
+use libcoap_sys::{coap_delete_oscore_conf, coap_new_oscore_conf, coap_new_str_const, coap_oscore_conf_t};
+
+use crate::error::OscoreConfigError;
+
+/// Represents an oscore config object which stores the underlying
+/// coap_oscore_conf_t C struct.
+pub struct OscoreConf {
+    raw_conf: *mut coap_oscore_conf_t,
+    pub(crate) initial_recipient: Option<String>,
+    drop: bool,
+}
+
+impl OscoreConf {
+    /// Creates a new OscoreConf.
+    ///
+    /// # Errors
+    /// Will return a [OscoreConfigError] if creating the oscore config fails (most likely due to
+    /// invalid oscore config bytes provided).
+    pub fn new(
+        seq_initial: u64,
+        oscore_conf_bytes: &[u8],
+        save_seq_num_func: extern "C" fn(seq_num: u64, _param: *mut c_void) -> i32,
```

Comment:
Ideally, we should allow `save_seq_num_func` to be a) a regular Rust function (or even a lambda) and b) have a parameter passed to it (just like with the resource handlers).

For now, this is fine. I'll create a follow up issue and link to this comment for future reference.